test/mpi/distributed/preconditioner/schwarz.cpp

```diff
@@ -0,0 +1,224 @@
+        auto dist_result =
+            dist_vec_type::create(ref, comm, global_size, local_size, nrhs);
+        dist_result->read_distributed(
+            gen_dense_data<typename dist_vec_type::value_type,
+                           global_index_type>(global_size),
+            row_part.get());
+        dist_b = gko::share(gko::clone(exec, dist_result));
+        dist_x = gko::share(gko::clone(exec, dist_result));
+        dist_x->fill(gko::zero<value_type>());
+
+        local_solver_factory =
+            local_prec_type::build().with_max_block_size(1u).on(exec);
+    }
+
+    void SetUp() override { ASSERT_EQ(comm.size(), 3); }
+
+
+    template <typename ValueType, typename IndexType>
+    gko::matrix_data<ValueType, IndexType> gen_dense_data(gko::dim<2> size)
+    {
+        return {
+            size,
+            std::normal_distribution<gko::remove_complex<ValueType>>(0.0, 1.0),
+            engine};
+    }
+
+    template <typename M1, typename DistVecType>
+    void assert_residual_near(const std::shared_ptr<M1>& mtx,
+                              const std::shared_ptr<DistVecType>& x,
+                              const std::shared_ptr<DistVecType>& b,
+                              double tolerance)
+    {
+        auto one =
+            gko::initialize<local_vec_type>({gko::one<value_type>()}, exec);
+        auto neg_one =
+            gko::initialize<local_vec_type>({-gko::one<value_type>()}, exec);
+        auto norm = DistVecType::local_vector_type::absolute_type::create(
+            ref, gko::dim<2>{1, b->get_size()[1]});
+        auto dist_res = gko::clone(b);
+        auto b_norm = DistVecType::local_vector_type::absolute_type::create(
+            ref, gko::dim<2>{1, b->get_size()[1]});
+        b->compute_norm2(b_norm.get());
+        mtx->apply(neg_one.get(), x.get(), one.get(), dist_res.get());
+        dist_res->compute_norm2(norm.get());
+
+        for (int i = 0; i < norm->get_num_stored_elements(); ++i) {
+            EXPECT_LE(norm->at(i) / b_norm->at(i), tolerance);
+        }
+    }
+
+
+    gko::dim<2> size;
+    std::shared_ptr<Partition> row_part;
+
+    gko::matrix_data<value_type, global_index_type> mat_input;
+
+    std::shared_ptr<dist_mtx_type> dist_mat;
+    std::shared_ptr<dist_vec_type> dist_b;
+    std::shared_ptr<dist_vec_type> dist_x;
+    std::shared_ptr<gko::LinOpFactory> solver_factory;
+    std::shared_ptr<gko::LinOpFactory> local_solver_factory;
+
+    std::default_random_engine engine;
+};
+
+TYPED_TEST_SUITE(SchwarzPreconditioner, gko::test::ValueLocalGlobalIndexTypes,
+                 TupleTypenameNameGenerator);
+
+
+TYPED_TEST(SchwarzPreconditioner, CanApplyPreconditionedSolver)
+{
+    using value_type = typename TestFixture::value_type;
+    using csr = typename TestFixture::local_matrix_type;
+    using cg = typename TestFixture::solver_type;
+    using prec = typename TestFixture::dist_prec_type;
+    constexpr double tolerance = 1e-3;
```

Comment:
the tolerance is quite low here? it's not converged at all?

Reply:
I think we do the same for our solver tests as well. So the tolerances are similar as it is there.

Reply:
why do we only set 1e-3 for mpi solver?

Reply:
I think we take general matrices and make them hpd. So it might be that we they are not well behaved for all solvers. I think those common tests need to be checked and we need to check the tolerances there.